### PR TITLE
Pin/Unpin Logic

### PR DESCRIPTION
**Pull request by Velin92**
Requires an updated SDK containing the following PR: https://github.com/matrix-org/matrix-rust-sdk/pull/3750

fixes https://github.com/element-hq/element-x-ios/issues/3058

https://github.com/user-attachments/assets/b2e86389-5f16-40fc-a1c3-4b1476c2dd09

### CODE REVIEW

File: ElementX/Sources/Screens/RoomScreen/View/ItemMenu/TimelineItemMenuActionProvider.swift [hunk 1]
```diff
@@ -21,6 +21,7 @@ struct TimelineItemMenuActionProvider {
     let canCurrentUserRedactSelf: Bool
     let canCurrentUserRedactOthers: Bool
     let canCurrentUserPin: Bool
+    let pinnedEvents: Set<String>
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/View/ItemMenu/TimelineItemMenuActionProvider.swift:24]: Might be slightly clearer if this was either `pinnedEventIDs` or `: Set<TimelineItemIdentifier>`, not sure it matters which :)

File: ElementX/Sources/Screens/RoomScreen/View/RoomScreen.swift [hunk 2]
```diff
@@ -109,10 +110,10 @@ struct RoomScreen: View {
     }
     
     private var pinnedItemsBanner: some View {
-        PinnedItemsBannerView(pinIndex: context.viewState.currentPinIndex,
-                              pinsCount: context.viewState.pinnedItems.count,
-                              pinContent: context.viewState.selectedPinContent,
-                              onMainButtonTap: { context.send(viewAction: .nextPin) },
+        PinnedItemsBannerView(pinIndex: context.viewState.pinnedEventsState.selectedPinIndex,
+                              pinsCount: context.viewState.pinnedEventsState.pinnedEvents.count,
+                              pinContent: context.viewState.pinnedEventsState.selectedPinContent,
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/View/RoomScreen.swift:115]: Would it make sense to pass this the pinnedEventsState as a single parameter now?

File: ElementX/Sources/Services/Room/RoomProxyProtocol.swift [hunk 3]
```diff
@@ -38,6 +38,7 @@ protocol RoomProxyProtocol {
     var isSpace: Bool { get }
     var isEncrypted: Bool { get }
     var isFavourite: Bool { get async }
+    var pinnedEvents: [String] { get async }
```
Comment (pixlwave) [inline ElementX/Sources/Services/Room/RoomProxyProtocol.swift:41]: Same here I guess.

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 4]
```diff
@@ -202,6 +202,24 @@ final class TimelineProxy: TimelineProxyProtocol {
         }
     }
     
+    func pin(eventID: String) async -> Result<Bool, TimelineProxyError> {
+        do {
+            return try await .success(timeline.pinEvent(eventIdStr: eventID))
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:207]: Lol at the `Str` suffix, we should send the Rust folks a few more pennies so they can fill in the missing characters 🤭

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift [hunk 5]
```diff
@@ -404,12 +407,17 @@ class RoomScreenViewModel: RoomScreenViewModelType, RoomScreenViewModelProtocol
         roomProxy
             .actionsPublisher
             .filter { $0 == .roomInfoUpdate }
-            .throttle(for: .seconds(1), scheduler: DispatchQueue.main, latest: true)
+            .receive(on: DispatchQueue.main)
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:410]: What happened to the throttle, I would imagine this was needed for a specific reason no?

File: ElementX/Resources/Localizations/en.lproj/Localizable.strings [hunk 6]
```diff
@@ -103,6 +103,7 @@
 "action_yes" = "Yes";
 "action.load_more" = "Load more";
 "action.pin" = "Pin";
+"action.unpin" = "Unpin";
```
Comment (pixlwave) [inline ElementX/Resources/Localizations/en.lproj/Localizable.strings:106]: Think these are missing the nested option on Localazy btw:
Screenshot 2024-07-24 at 6 13 36 PM

File: ElementX/Sources/Screens/RoomScreen/RoomScreenModels.swift [hunk 7]
```diff
@@ -304,3 +294,40 @@ enum ScrollDirection: Equatable {
...
+                selectedPin = pinnedEvents.first
+            } else if pinnedEvents.isEmpty {
+                selectedPin = nil
+            } else if let selectedPin, !pinnedEvents.contains(selectedPin) {
+                self.selectedPin = pinnedEvents.first
+            }
+        }
+    }
+    
+    var selectedPin: String?
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/RoomScreenModels.swift:312]: And the same here, `selectedPinEventID` would be clearer :)

**Review by pixlwave — APPROVED** [review]
🤩

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift (hunk 5 above)
Comment (Velin92) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:410]: Yeah so regarding the throttle... it makes the pinning banner updates slower... maybe what I could is remove it and do an if check on every value, and only change them if they are not equal to the new value, WDYT?

**Pull request merged**